### PR TITLE
Refactor to be customizable.

### PR DESCRIPTION
**Pull request by whatthejeff**
@sebastianbergmann This is an attempt to make things more modular / customizable and easier to use in other libs which have custom visualization needs. Thoughts?

### CODE REVIEW

**Comment by sebastianbergmann** [comment]
Somehow we / I forgot to merge this and now it conflicts with `master`. Can you rebase, @whatthejeff?

**Comment by whatthejeff** [comment]
Not a problem. Give me a couple of days.

**Comment by stof** [comment]
@whatthejeff any news of the rebase ?

File: src/SplObjectStorageExporter.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        $array = parent::toArray($value);
+
+        // Remove HHVM internal representations
+        if (property_exists('\SplObjectStorage', '__storage')) {
+          unset($array['__storage']);
+        } elseif (property_exists('\SplObjectStorage', 'storage')) {
+          unset($array['storage']);
+        }
+        if (property_exists('\SplObjectStorage', '__key')) {
+          unset($array['__key']);
```
Comment (stof) [inline src/SplObjectStorageExporter.php]: wrong indentation

File: src/SplObjectStorageExporter.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        // Some internal classes like SplObjectStorage don't work with our
+        // normal method of converting objects to arrays.
```
Comment (stof) [inline src/SplObjectStorageExporter.php]: this comment looks like it is in the wrong place

File: src/StringExporter.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * Recursively exports a value as a string.
+     *
+     * @param  mixed                              $value       The value to export
+     * @param  integer                            $indentation The indentation level of the 2nd+ line
+     * @param  SebastianBergmann\Exporter\Context $processed   Contains all objects and arrays that have previously been rendered
+     * @return string
+     * @see    SebastianBergmann\Exporter\Exporter::export
+     */
+    protected function recursiveExport(&$value, $indentation, $processed = null)
```
Comment (stof) [inline src/StringExporter.php]: implementing this as ``recursiveExport`` looks wrong to me, as it is not recursive at all). IMO, this should be implemented as ``export`` directly (without extending BaseExporter, as you don't really need any code from it, but implementing an interface used by any of the exporters)

File: src/Exporter.php [hunk 4]
```diff
@@ -27,237 +27,76 @@
  * @license    http://www.opensource.org/licenses/BSD-3-Clause  The BSD 3-Clause License
  * @link       https://github.com/sebastianbergmann/exporter
  */
-class Exporter
+class Exporter extends BaseExporter
```
Comment (stof) [inline src/Exporter.php]: why does it extend BaseExporter ? It does not use any code from it

File: src/BaseExporter.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        return $this->recursiveExport($value, $indentation);
+    }
+
+    /**
+     * Recursively exports a value as a string.
+     *
+     * @param  mixed                              $value       The value to export
+     * @param  integer                            $indentation The indentation level of the 2nd+ line
+     * @param  SebastianBergmann\Exporter\Context $processed   Contains all objects and arrays that have previously been rendered
```
Comment (stof) [inline src/BaseExporter.php]: IMO, this argument should be typehinted

File: src/BaseExporter.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        $this->factory = $factory;
+    }
+
+    /**
+     * Returns whether the exporter can export a given value.
+     *
+     * @param  mixed   $value The value to export.
+     * @return boolean
+     */
+    public function accepts($value)
```
Comment (stof) [inline src/BaseExporter.php]: IMO, you should force sub-classes to define explicitly what they accept, to avoid mistakes

**Pull request closed without merging**